pkg/providers/tinkerbell/reconciler/reconciler.go

```diff
@@ -109,17 +119,54 @@ func (r *Reconciler) ValidateClusterSpec(ctx context.Context, log logr.Logger, t
 	return controller.Result{}, nil
 }
 
-// ReconcileControlPlane applies the control plane CAPI objects to the cluster.
-func (r *Reconciler) ReconcileControlPlane(ctx context.Context, log logr.Logger, tinkerbellScope *Scope) (controller.Result, error) {
+// DetectOperationAndGenerateSpec generates control plane and workers spec, then detect create or upgrade type.
+func (r *Reconciler) DetectOperationAndGenerateSpec(ctx context.Context, log logr.Logger, tinkerbellScope *Scope) (controller.Result, error) {
 	spec := tinkerbellScope.ClusterSpec
-	log = log.WithValues("phase", "reconcileControlPlane")
-	log.Info("Applying control plane CAPI objects")
+	log = log.WithValues("phase", "DetectOperation")
+	log.Info("Detecting operation type")
+
 	cp, err := tinkerbell.ControlPlaneSpec(ctx, log, clientutil.NewKubeClient(r.client), spec)
 	if err != nil {
 		return controller.Result{}, err
 	}
+	tinkerbellScope.controlPlane = cp
+
+	w, err := tinkerbell.WorkersSpec(ctx, log, clientutil.NewKubeClient(r.client), spec)
+	if err != nil {
+		return controller.Result{}, errors.Wrap(err, "generating workers spec")
+	}
+	tinkerbellScope.workers = w
+
+	err = detectOperation(ctx, log, tinkerbellScope)
+	if err != nil && apierrors.IsNotFound(err) {
+		return controller.Result{}, errors.Wrap(err, "Detecting operation type")
+	}
+
+	// omit machine template on scale update
```

Comment:
why do they need to be nullified? can't then just have the same value as before?

Reply:
My understanding is we don't want to create new machine template on scale. @pokearu may have more context here.

Reply:
If I recall correctly its simply because the existing template is sufficient to perform the scale up. 

The generation logic needs rethinking (separate to these changes) because generate and omit under specific circumstances feels convoluted and error prone.